README.md

```diff
@@ -1,229 +1,894 @@
+
+### Default stack trace mode
+
+As mentioned above, by default `retsnoop` is capturing a stack trace, with the
+deepest nested function calls that satisfy conditions (i.e., if no custom error
+filtering is specified, `retsnoop` will try to capture stack trace of a deepest
+function call chain resulting in error return). See
+[first example](#failed-per-cpu-bpf-array-map-case) for and example output in
+this mode, and how to interpret it.
+
+This mode is currently always on and can't be disabled. Other modes (function
+calls trace and LBR, described below) are complementary to default stack trace
+mode (and each other as well).
+
+### Function calls trace mode
+
+With `-T` (`--trace`) function calls trace mode is enabled. In this mode,
+`retsnoop` will keep track of a detailed sequence of calls between each entry
+and non-entry function. Just like stack trace mode, this recording is activated
+only upon hitting an entry function and also satisfying any of the additional
+filtering conditions.
+
+As mentioned above, this mode is complimenting default stack trace mode, and,
+if enabled, LBR mode. It provides a different view of captured function call
+sequence. Given it might be quite verbose and expensive to record, depending on
+specific workload and a set of functions of interest, it requires explicit
+opt-in with `-T` argument.
+
+This mode is perfect for understanding kernel behavior in details, especially
+unfamiliar parts of it. See [second example](#tracing-bpf-verification-flow)
+for details on understanding output format and how it can be interpreted.
+
+
+### LBR (Last Branch Records) mode
+
+[LBR](https://lwn.net/Articles/680985/) (Last Branch Records) is Intel CPUs
+feature that allows to instruct CPU to record last N calls/returns/jumps (it's
+configurable) constantly with no overhead (where N depends on generation of CPU
+and is typically in 16-32 range), and in recent enough Linux kernel (v5.16+)
+it's possible to capture such LBRs from BPF program in ad-hoc fashion, which
+can be utilized by `retsnoop` in LBR mode. Non-Intel CPUs might have similar
+set of features, abstracted away somewhat by kernel's perf subsystem, so you
+don't necessarily need Intel CPUs to take advantage of this with `retsnoop`.
+
+So when is LBR mode useful? There are a few typical scenarios.
+
+One of them is when you are investigating some generic error being returned
+from kernel. And such error can be returned in various different cases and it's
+not clear which one it is. This is often the case with `bpf()` syscall, where
+errors like `-EINVAL` can be returned due to dozens of various error
+conditions, and a bunch of such conditions could be checked within the same
+function. Debugging exactly which error condition is hit can maddening at
+times. LBR mode allows to get insight *into which exact if condition within the
+function* returned an error. See
+[third example](#peering-deep-into-functions-with-lbr) for exactly such use
+case.
```

Comment:
not a review comment, but god I love this feature lol

Reply:
I know, me too! :) At some point I almost gave up on this, but Song managed to somehow allow this ad-hoc LBR capture from BPF.